add support for legacy/non-gki kernels

maxsteeel opened the pull request:
I'm reopening this PR in legacy branch to make it easier to add it as a non-gki branch

maxsteeel committed:
kernel: add support for legacy/non-gki kernels

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

maxsteeel committed:
kernel: fix compilation in non-gki kernels such as 5.4.210

Thanks @A7mdwassa for reporting bugs and @rsuntk for reference.

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

rifsxd committed:
manager: filter ksuapp in superuser screen
this will allow showing kernelsu app info in webui

Signed-off-by: KOWX712 <leecc0503@gmail.com>

maxsteeel committed:
kernel: "fix undefined symbol: disable_seccomp"

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

rifsxd commented:
good commit every ngki code to legacy branch, also whats the current situation about ngki support? is it only test on gki 1 4.19 - 5.4? what about below that like 4.4 - 4.14? debug/test it and keep me updated about this PR

rifsxd commented:
and when its stable and working on said tests and in the end let me know when its ready

rifsxd commented:
i will test the end commit

neonmodder123 commented:
Sadly, this PR fixes some errors but not all... I am building for kernel 5.4.254.
Logs:
`-- KernelSU-Next version: 32860
-- KernelSU-Next tag: v1.1.1
-- KernelSU-Next Manager signature size: 0x3e6
-- KernelSU-Next Manager signature hash: 79e590113c4c4c0c222978e413a5faa801666957b1212a328e46c00c69821bf7
  CC      drivers/kernelsu/ksu.o
  CC      drivers/iommu/iommu-sysfs.o
  CC      lib/rbtree.o
  CC      lib/seq_buf.o
  CC      drivers/kernelsu/allowlist.o
  CC      drivers/gpu/msm/adreno_a5xx_preempt.o
  CC      lib/sha1.o
  CC      techpack/video/msm/vidc/hfi_iris2.o
  CC      lib/show_mem.o
  CC      fs/coredump.o
  CC      lib/siphash.o
  CC      drivers/gpu/drm/drm_blend.o
  CC      drivers/irqchip/irq-gic-common.o
  CC      drivers/gpu/drm/drm_encoder.o
  CC      drivers/iommu/dma-iommu.o
  CC      drivers/kernelsu/app_profile.o
  CC      net/ipv4/ip_fragment.o
  CC      lib/string.o
  CC      drivers/irqchip/irq-gic-v2m.o
  CC      drivers/iommu/msm_dma_iommu_mapping.o
../drivers/kernelsu/app_profile.c:71:31: error: no member named 'filter_count' in 'struct seccomp'
        atomic_set(&current->seccomp.filter_count, 0);
                    ~~~~~~~~~~~~~~~~ ^
1 error generated.
make[3]: *** [../scripts/Makefile.build:288: drivers/kernelsu/app_profile.o] Error 1
make[2]: *** [../scripts/Makefile.build:558: drivers/kernelsu] Error 2
make[2]: *** Waiting for unfinished jobs....
`

maxsteeel committed:
kernel: handle seccomp_filter_release compat

Signed-off-by: TwinbornPlate75 <3342733415@qq.com>
Signed-off-by: fc5b87cf <rissu.ntk@gmail.com>
Co-authored-by: TwinbornPlate75 <3342733415@qq.com>
Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

maxsteeel commented:
@neonmodder123 try with this new commit

maxsteeel committed:
kernel: implicit declaration of function 'strncpy_from_user_nofault'

Declaration from sucompat.c is passed to sucompat.h for compatibility.

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

zenwav commented:
Encountered this error while building Kernel 4.19.246
```
../drivers/kernelsu/pkg_observer.c:50:18: error: incompatible function pointer types initializing 'int (*)(struct fsnotify_group *, struct inode *, u32, const void *, int, const unsigned char *, u32, struct fsnotify_iter_info *)' (aka 'int (*)(struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const unsigned char *, unsigned int, struct fsnotify_iter_info *)') with an expression of type 'int (struct fsnotify_group *, struct inode *, u32, const void *, int, const struct qstr *, u32, struct fsnotify_iter_info *)' (aka 'int (struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const struct qstr *, unsigned int, struct fsnotify_iter_info *)') [-Wincompatible-function-pointer-types]
   50 |         .handle_event = ksu_handle_event,
      |                         ^~~~~~~~~~~~~~~~
1 error generated.
make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/pkg_observer.o] Error 1
make[2]: *** [../scripts/Makefile.build:521: drivers/kernelsu] Error 2
make[1]: *** [/home/ubuntu/n0kernel/Makefile:1236: drivers] Error 2
make[1]: *** Waiting for unfinished jobs....
```

rifsxd commented:
> Encountered this error while building Kernel 4.19.246
> 
> ```
> ../drivers/kernelsu/pkg_observer.c:50:18: error: incompatible function pointer types initializing 'int (*)(struct fsnotify_group *, struct inode *, u32, const void *, int, const unsigned char *, u32, struct fsnotify_iter_info *)' (aka 'int (*)(struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const unsigned char *, unsigned int, struct fsnotify_iter_info *)') with an expression of type 'int (struct fsnotify_group *, struct inode *, u32, const void *, int, const struct qstr *, u32, struct fsnotify_iter_info *)' (aka 'int (struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const struct qstr *, unsigned int, struct fsnotify_iter_info *)') [-Wincompatible-function-pointer-types]
>    50 |         .handle_event = ksu_handle_event,
>       |                         ^~~~~~~~~~~~~~~~
> 1 error generated.
> make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/pkg_observer.o] Error 1
> make[2]: *** [../scripts/Makefile.build:521: drivers/kernelsu] Error 2
> make[1]: *** [/home/ubuntu/n0kernel/Makefile:1236: drivers] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```

ksu_handle_event is no longer needed in 3.0

rifsxd commented:
> > Encountered this error while building Kernel 4.19.246
> > ```
> > ../drivers/kernelsu/pkg_observer.c:50:18: error: incompatible function pointer types initializing 'int (*)(struct fsnotify_group *, struct inode *, u32, const void *, int, const unsigned char *, u32, struct fsnotify_iter_info *)' (aka 'int (*)(struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const unsigned char *, unsigned int, struct fsnotify_iter_info *)') with an expression of type 'int (struct fsnotify_group *, struct inode *, u32, const void *, int, const struct qstr *, u32, struct fsnotify_iter_info *)' (aka 'int (struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const struct qstr *, unsigned int, struct fsnotify_iter_info *)') [-Wincompatible-function-pointer-types]
> >    50 |         .handle_event = ksu_handle_event,
> >       |                         ^~~~~~~~~~~~~~~~
> > 1 error generated.
> > make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/pkg_observer.o] Error 1
> > make[2]: *** [../scripts/Makefile.build:521: drivers/kernelsu] Error 2
> > make[1]: *** [/home/ubuntu/n0kernel/Makefile:1236: drivers] Error 2
> > make[1]: *** Waiting for unfinished jobs....
> > ```
> 
> ksu_handle_event is no longer needed in 3.0

Matter of fact the whole pkg_observer is dropped after 3.0

rifsxd commented:
> Encountered this error while building Kernel 4.19.246
> 
> ```
> ../drivers/kernelsu/pkg_observer.c:50:18: error: incompatible function pointer types initializing 'int (*)(struct fsnotify_group *, struct inode *, u32, const void *, int, const unsigned char *, u32, struct fsnotify_iter_info *)' (aka 'int (*)(struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const unsigned char *, unsigned int, struct fsnotify_iter_info *)') with an expression of type 'int (struct fsnotify_group *, struct inode *, u32, const void *, int, const struct qstr *, u32, struct fsnotify_iter_info *)' (aka 'int (struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const struct qstr *, unsigned int, struct fsnotify_iter_info *)') [-Wincompatible-function-pointer-types]
>    50 |         .handle_event = ksu_handle_event,
>       |                         ^~~~~~~~~~~~~~~~
> 1 error generated.
> make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/pkg_observer.o] Error 1
> make[2]: *** [../scripts/Makefile.build:521: drivers/kernelsu] Error 2
> make[1]: *** [/home/ubuntu/n0kernel/Makefile:1236: drivers] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```

The code should be handle_inode_event

maxsteeel committed:
kernel: fix ksu_handle_inode_event implementation for non-gki kernels

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

neonmodder123 commented:
Thank you so much @maxsteeel ! My 5.4 kernel built successfully. Will see if KernelSU Next works properly :)

maxsteeel commented:
Now I will start fixing compilation errors in a 4.14 kernel :)

maxsteeel commented:
> Encountered this error while building Kernel 4.19.246
> 
> ```
> ../drivers/kernelsu/pkg_observer.c:50:18: error: incompatible function pointer types initializing 'int (*)(struct fsnotify_group *, struct inode *, u32, const void *, int, const unsigned char *, u32, struct fsnotify_iter_info *)' (aka 'int (*)(struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const unsigned char *, unsigned int, struct fsnotify_iter_info *)') with an expression of type 'int (struct fsnotify_group *, struct inode *, u32, const void *, int, const struct qstr *, u32, struct fsnotify_iter_info *)' (aka 'int (struct fsnotify_group *, struct inode *, unsigned int, const void *, int, const struct qstr *, unsigned int, struct fsnotify_iter_info *)') [-Wincompatible-function-pointer-types]
>    50 |         .handle_event = ksu_handle_event,
>       |                         ^~~~~~~~~~~~~~~~
> 1 error generated.
> make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/pkg_observer.o] Error 1
> make[2]: *** [../scripts/Makefile.build:521: drivers/kernelsu] Error 2
> make[1]: *** [/home/ubuntu/n0kernel/Makefile:1236: drivers] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```

try with latest commit here

rifsxd commented:
> Now I will start fixing compilation errors in a 4.14 kernel :)

good luck

rifsxd commented:
watching closely

maxsteeel committed:
kernel: pkg_observer: fix build in non-gki kernels

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

zenwav commented:
> try with latest commit here
```
../drivers/kernelsu/supercalls.c:389:38: error: implicit declaration of function 'selinux_inode' [-Werror,-Wimplicit-function-declaration]
  389 |         struct inode_security_struct *sec = selinux_inode(wrapper_inode);
      |                                             ^
1 error generated.
make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/supercalls.o] Error 1
```

maxsteeel committed:
kernel: add compatibility with kernels 4.14 and above

The code was partially restructured to support non-gki kernels such as 4.14. Previous commits were also cleaned up.

I have only been tested on kernels 4.14.356 and 5.4.254.

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

maxsteeel committed:
kernel: add compatibility with kernels 4.14 and above

The code was partially restructured to support non-gki kernels such as 4.14. Previous commits were also cleaned up.

I have only been tested on kernels 4.14.356 and 5.4.254.

Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

maxsteeel commented:
> > try with latest commit here
> 
> ```
> ../drivers/kernelsu/supercalls.c:389:38: error: implicit declaration of function 'selinux_inode' [-Werror,-Wimplicit-function-declaration]
>   389 |         struct inode_security_struct *sec = selinux_inode(wrapper_inode);
>       |                                             ^
> 1 error generated.
> make[3]: *** [../scripts/Makefile.build:307: drivers/kernelsu/supercalls.o] Error 1
> ```

try latest commit

Sorayukii commented:
Tested for 4.9?

maxsteeel commented:
> Tested for 4.9?

I haven't tried it, but you can try it, and let me know if you encounter any compilation error

NTeditor commented:
Kernel version: 5.4.254
```
-- KDIR:
-- MDIR: /kernel-source/KernelSU-Next/kernel
-- KernelSU-Next version: 32871
-- KernelSU-Next tag: v3.0.0
-- KernelSU-Next Manager signature size: 0x3e6
-- KernelSU-Next Manager signature hash: 79e590113c4c4c0c222978e413a5faa801666957b1212a328e46c00c69821bf7
-- KernelSU: CONFIG_KSU_SYSCALL_HOOK
  CC      drivers/kernelsu/ksud.o
../drivers/kernelsu/ksud.c:80:27: error: unused variable 'stop_vfs_read_work' [-Werror,-Wunused-variable]
static struct work_struct stop_vfs_read_work;
                          ^
../drivers/kernelsu/ksud.c:81:27: error: unused variable 'stop_execve_hook_work' [-Werror,-Wunused-variable]
static struct work_struct stop_execve_hook_work;
                          ^
../drivers/kernelsu/ksud.c:82:27: error: unused variable 'stop_input_hook_work' [-Werror,-Wunused-variable]
static struct work_struct stop_input_hook_work;
                          ^
3 errors generated.
make[3]: *** [../scripts/Makefile.build:288: drivers/kernelsu/ksud.o] Error 1
make[2]: *** [../scripts/Makefile.build:560: drivers/kernelsu] Error 2
make[1]: *** [/kernel-source/Makefile:1847: drivers] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/kernel-source/out'
make: *** [Makefile:186: sub-make] Error 2
```

With this patch, compilation is successful:
```diff
diff --git a/kernel/ksud.c b/kernel/ksud.c
index 595370a9..b9978adc 100644
--- a/kernel/ksud.c
+++ b/kernel/ksud.c
@@ -77,9 +77,9 @@ static void stop_execve_hook();
 static void stop_input_hook();

 #ifdef CONFIG_KSU_SYSCALL_HOOK
-static struct work_struct stop_vfs_read_work;
-static struct work_struct stop_execve_hook_work;
-static struct work_struct stop_input_hook_work;
+static struct work_struct __maybe_unused stop_vfs_read_work;
+static struct work_struct __maybe_unused stop_execve_hook_work;
+static struct work_struct __maybe_unused stop_input_hook_work;
 #else
 bool ksu_vfs_read_hook __read_mostly = true;
 bool ksu_execveat_hook __read_mostly = true;
@@ -577,4 +577,4 @@ void ksu_ksud_exit(void)
 #ifdef CONFIG_KSU_SYSCALL_HOOK
        kp_handle_ksud_exit();
 #endif
-}
\ No newline at end of file
+}
```

rsuntk commented:
Some of these codes are taken from RKSU, and no proper authorship were given. 👀

rifsxd commented:
> Some of these codes are taken from RKSU, and no proper authorship were given. 👀

Yeah I think he mentioned somewhere in telegram but no actual signing...

@maxsteeel you better sign the commits

zenwav commented:
> try latest commit
```
  MODPOST vmlinux.o
ld.lld: error: undefined symbol: ksu_handle_sys_reboot
>>> referenced by ld-temp.o
>>>               vmlinux.o:(__arm64_sys_reboot)
```

rifsxd commented:
this PR is expanding beyond the intended scope, we should be focusing on manual hooks first while keeping krpobes from original code as it is, but now you are bringing in unnecessary hybrid of krpobes and tracepoint hooks, we dont need tracepoints and krpobes focused hook, tbh no tracepoint hooks at all, first need a manual hook working state then we can bring back traditional krpobes from dev branch to pass the previous KSU_KRPOBES_HOOk config


also you are missing the ksu_handle_sys_reboot entirely

rifsxd commented:
sorry for inconvenience but i had to get rid of an unclean and no tagged branch so i deleted and created new branch matching stable branch

maxsteeel committed:
kernel: restructure code for kprobes and manual hooks support

Based on RKSU, this commit adds support for kernels without kprobes / with manual hooks, while maintaining support for GKI1 kernels with working kprobes.

Credits to @rsuntk, RKSU maintainer.

Co-authored-by: rsuntk <rissu.ntk@gmail.com>
Signed-off-by: maxsteeel <109047395+maxsteeel@users.noreply.github.com>

maxsteeel commented:
The latest commit is working fully for me with manual hooks v1.6 without kprobes on 5.4 and 4.14 kernels (those I testing). Need testing on older kernels

sidex15 commented:
<img width="500" alt="image" src="https://github.com/user-attachments/assets/d15e931f-3108-4257-aae5-1e63c8cd4645" />

Tested and working fine now!
